pkg/daemon/pinned_image_set.go

```diff
@@ -0,0 +1,1273 @@
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					continue
+				}
+				klog.Errorf("Error getting pinned image set: %v", err)
+				return nil, nil, err
+			}
+
+			imageSetConfig := machineconfigurationalphav1.MachineConfigNodeStatusPinnedImageSet().
+				WithName(imageSet.Name).
+				WithDesiredGeneration(int32(imageSet.GetGeneration()))
+			if statusErr != nil {
+				imageSetConfig.LastFailedGeneration = ptr.To(int32(imageSet.GetGeneration()))
+				imageSetConfig.LastFailedGenerationErrors = []string{statusErr.Error()}
+			} else {
+				imageSetConfig.CurrentGeneration = ptr.To(int32(imageSet.GetGeneration()))
+			}
+
+			imageSetApplyConfigs = append(imageSetApplyConfigs, imageSetConfig)
+		}
+	}
+	return imageSetApplyConfigs, mcnImageSets, nil
+}
+
+// getWorkerCount returns the number of workers to use for prefetching images.
+func (p *PinnedImageSetManager) getWorkerCount() (int, error) {
+	node, err := p.nodeLister.Get(p.nodeName)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get node %q: %w", p.nodeName, err)
+	}
+
+	// default to 5 workers
+	workerCount := defaultPrefetchWorkers
+
+	// master nodes are a special case and should have a concurrency of 1 to
+	// mitigate I/O contention with the control plane.
+	if _, ok := node.ObjectMeta.Labels["node-role.kubernetes.io/master"]; ok {
+		workerCount = defaultControlPlaneWorkers
+	}
+
+	return workerCount, nil
+}
+
+func (p *PinnedImageSetManager) resetWorkload(cancelFn context.CancelFunc) {
+	p.mu.Lock()
+	if p.cancelFn != nil {
+		p.cancelFn()
+	}
+	p.cancelFn = cancelFn
+	p.mu.Unlock()
+}
+
+func (p *PinnedImageSetManager) cancelWorkload(reason string) {
+	klog.Infof("Cancelling workload: %s", reason)
+	p.mu.Lock()
+	if p.cancelFn != nil {
+		p.cancelFn()
+		p.cancelFn = nil
+	}
+	p.mu.Unlock()
+}
+
+// prefetchWorker is a worker that pulls images from the container runtime.
+func (p *PinnedImageSetManager) prefetchWorker(ctx context.Context) {
+	for task := range p.prefetchCh {
+		if err := ensurePullImage(ctx, p.criClient, p.backoff, task.image, task.auth); err != nil {
+			task.monitor.Error(err)
+			klog.Warningf("failed to prefetch image %q: %v", task.image, err)
+		}
+		task.monitor.Done()
+
+		cachedImage, ok := p.cache.Get(task.image)
+		if ok {
+			imageInfo := cachedImage.(imageInfo)
+			imageInfo.Pulled = true
+			p.cache.Add(task.image, imageInfo)
+		} else {
+			p.cache.Add(task.image, imageInfo{Name: task.image, Pulled: true})
+		}
+
+		// throttle prefetching to avoid overloading the file system
+		select {
+		case <-time.After(defaultPrefetchThrottleDuration):
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// scheduleWork schedules the prefetch work for the images and collects the first error encountered.
+func (p *PinnedImageSetManager) scheduleWork(ctx context.Context, prefetchCh chan prefetch, registryAuth *registryAuth, prefetchImages []mcfgv1alpha1.PinnedImageRef, monitor *prefetchMonitor) error {
```

Comment:
this can't get called with 0 images to pull right? just double checking

Reply:
will make this more explicit.